File: HackerNews/Views/StoryScreen.swift
import SwiftUI

/// A story and its discussion.
struct StoryScreen: View {
    @StateObject private var store: ThreadStore
    @Binding var path: NavigationPath

    @EnvironmentObject private var settings: AppSettings
    @EnvironmentObject private var library: LibraryStore
    @EnvironmentObject private var opener: LinkOpener

    init(story: Item, path: Binding<NavigationPath>) {
        _store = StateObject(wrappedValue: ThreadStore(story: story))
        _path = path
    }

    var body: some View {
        List {
            Section {
                StoryHeader(story: store.story, path: $path)
                    .listRowInsets(EdgeInsets(top: 12, leading: 16, bottom: 12, trailing: 16))
                    .listRowSeparator(.hidden)
            }

            Section {
                switch store.phase {
                case .loading where store.nodes.isEmpty:
                    LoadingStateView(label: "Loading discussion…")
                        .frame(height: 160)
                        .listRowSeparator(.hidden)
                case .failed(let message):
                    ErrorStateView(message: message) {
                        store.retry(autoCollapseDeep: settings.autoCollapseDeepThreads)
                    }
                    .frame(height: 220)
                    .listRowSeparator(.hidden)
                default:
                    if store.nodes.isEmpty {
                        EmptyStateView(
                            systemImage: "bubble.left",
                            title: "No comments yet",
                            message: "Be the first — open this story on Hacker News to reply."
                        )
                        .frame(height: 200)
                        .listRowSeparator(.hidden)
                    } else {
                        ForEach(store.visibleNodes) { node in
                            CommentRow(
                                node: node,
                                isCollapsed: store.isCollapsed(node.id),
                                isOriginalPoster: node.author != nil && node.author == store.story.by,
                                path: $path
                            ) {
                                withAnimation(.easeOut(duration: 0.18)) {
                                    store.toggleCollapse(node)
                                }
                                Haptics.tap()
                            }
                            .listRowInsets(
                                EdgeInsets(top: 8, leading: 12, bottom: 8, trailing: 14)
                            )
                        }
                    }
                }
            } header: {
                if !store.nodes.isEmpty {
                    commentsHeader
                }
            }
        }
        .listStyle(.plain)
        .navigationTitle(store.nodes.isEmpty ? "Discussion" : store.totalComments.pluralized("comment"))
        .navigationBarTitleDisplayMode(.inline)
        .toolbar { toolbarContent }
        .task { store.loadIfNeeded(autoCollapseDeep: settings.autoCollapseDeepThreads) }
        .refreshable { await store.refresh(autoCollapseDeep: settings.autoCollapseDeepThreads) }
        .onChange(of: store.story) { updated in
            library.refreshSaved(with: updated)
        }
    }

    private var commentsHeader: some View {
        HStack {
            Text(store.totalComments.pluralized("comment").uppercased())
                .font(.caption2.weight(.semibold))
                .foregroundColor(.secondary)
            Spacer()
            Button {
                withAnimation(.easeOut(duration: 0.18)) {
                    if store.isFullyCollapsed { store.expandAll() } else { store.collapseAll() }
                }
                Haptics.tap()
            } label: {
                Label(
                    store.isFullyCollapsed ? "Expand All" : "Collapse All",
                    systemImage: store.isFullyCollapsed
                        ? "arrow.down.left.and.arrow.up.right"
                        : "arrow.up.right.and.arrow.down.left"
                )
                .font(.caption2)
                .labelStyle(.titleAndIcon)
            }
            .buttonStyle(.plain)
            .foregroundColor(.hnOrange)
        }
    }

    @ToolbarContentBuilder
    private var toolbarContent: some ToolbarContent {
        ToolbarItem(placement: .navigationBarTrailing) {
            Menu {
                if let link = store.story.link {
                    Button {
                        opener.open(link)
                    } label: {
                        Label("Open Article", systemImage: "safari")
                    }
                    Button {
                        opener.openInApp(link, readerMode: true)
                    } label: {
                        Label("Open in Reader", systemImage: "doc.plaintext")
                    }
                }

                Button {
                    opener.open(store.story.hnURL)
                } label: {
                    Label("Open on Hacker News", systemImage: "network")
                }

                Button {
                    if library.toggleSave(store.story) { Haptics.success() } else { Haptics.tap() }
                } label: {
                    Label(
                        library.isSaved(store.story.id) ? "Remove from Saved" : "Save Story",
                        systemImage: library.isSaved(store.story.id) ? "bookmark.slash" : "bookmark"
                    )
                }

                Divider()

                ShareLink(item: store.story.primaryURL) {
                    Label("Share Link", systemImage: "square.and.arrow.up")
                }
                ShareLink(item: store.story.hnURL) {
                    Label("Share Discussion", systemImage: "square.and.arrow.up.on.square")
                }
            } label: {
                Image(systemName: "ellipsis.circle")
            }
        }
    }
}

/// The story card that sits above the comments.
private struct StoryHeader: View {
    let story: Item
    @Binding var path: NavigationPath

    @EnvironmentObject private var library: LibraryStore
    @EnvironmentObject private var opener: LinkOpener

    var body: some View {
        VStack(alignment: .leading, spacing: 12) {
            Button {
                opener.open(story.primaryURL)
            } label: {
                VStack(alignment: .leading, spacing: 8) {
                    Text(story.displayTitle)
                        .font(.title3.weight(.bold))
                        .foregroundColor(.primary)
                        .multilineTextAlignment(.leading)
                        .fixedSize(horizontal: false, vertical: true)

                    if let host = story.host {
                        HStack(spacing: 4) {
                            Image(systemName: "safari")
                                .font(.caption2)
                            Text(host)
                                .font(.footnote)
                            Image(systemName: "arrow.up.forward")
                                .font(.system(size: 9, weight: .semibold))
                        }
                        .foregroundColor(.hnOrange)
                    }
                }
                .frame(maxWidth: .infinity, alignment: .leading)
            }
            .buttonStyle(.plain)

            byline

            if let text = story.text, !text.isEmpty {
                RichText(html: text, font: .callout)
                    .padding(.top, 2)
            }

            actionBar
        }
    }

    private var byline: some View {
        HStack(spacing: 12) {
            if let score = story.score {
                MetaLabel(
                    systemImage: "arrowtriangle.up.fill",
                    text: score.pluralized("point"),
                    tint: .hnOrange
                )
            }
            Button {
                path.append(Route.user(story.author))
            } label: {
                MetaLabel(systemImage: "person", text: story.author)
            }
            .buttonStyle(.plain)

            if let time = story.time {
                MetaLabel(systemImage: "clock", text: time.shortAge)
            }
        }
    }

    private var actionBar: some View {
        HStack(spacing: 10) {
            if story.link != nil {
                ActionButton(title: "Read", systemImage: "doc.plaintext") {
                    opener.openInApp(story.primaryURL, readerMode: true)
                }
            }
            ActionButton(
                title: library.isSaved(story.id) ? "Saved" : "Save",
                systemImage: library.isSaved(story.id) ? "bookmark.fill" : "bookmark",
                isActive: library.isSaved(story.id)
            ) {
                if library.toggleSave(story) { Haptics.success() } else { Haptics.tap() }
            }
            ShareLink(item: story.primaryURL) {
                ActionButtonLabel(title: "Share", systemImage: "square.and.arrow.up", isActive: false)
            }
            Spacer(minLength: 0)
        }
    }
}

private struct ActionButton: View {
    let title: String
    let systemImage: String
    var isActive = false
    let action: () -> Void

    var body: some View {
        Button(action: action) {
            ActionButtonLabel(title: title, systemImage: systemImage, isActive: isActive)
        }
        .buttonStyle(.plain)
    }
}

private struct ActionButtonLabel: View {
    let title: String
    let systemImage: String
    let isActive: Bool

    var body: some View {
        HStack(spacing: 5) {
            Image(systemName: systemImage)
                .font(.caption)
            Text(title)
                .font(.caption.weight(.medium))
        }
        .padding(.horizontal, 11)
        .padding(.vertical, 7)
        .background(
            Capsule().fill(
                isActive
                    ? Color.hnOrange.opacity(0.16)
                    : Color(uiColor: .secondarySystemBackground)
            )
        )
        .foregroundColor(isActive ? .hnOrange : .primary)
    }
}

File: HackerNewsAndroid/app/src/main/java/com/nurtrino/hackernews/ui/StoryScreen.kt
package com.nurtrino.hackernews.ui

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.ArrowBack
import androidx.compose.material.icons.filled.List
import androidx.compose.material.icons.filled.MoreVert
import androidx.compose.material.icons.filled.Refresh
import androidx.compose.material.icons.filled.Warning
import androidx.compose.material3.DropdownMenu
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.material3.TopAppBar
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import com.nurtrino.hackernews.data.Library
import com.nurtrino.hackernews.data.Settings
import com.nurtrino.hackernews.data.ThreadViewModel
import com.nurtrino.hackernews.model.LoadPhase

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun StoryScreen(
    model: ThreadViewModel,
    settings: Settings,
    library: Library,
    onBack: () -> Unit,
    onOpenUser: (String) -> Unit,
) {
    val context = LocalContext.current
    var menuOpen by remember { mutableStateOf(false) }
    val story = model.story
    val visible = model.visibleNodes

    LaunchedEffect(Unit) { model.loadIfNeeded(settings.autoCollapseDeepThreads) }
    LaunchedEffect(story) { library.refreshSaved(story) }

    val openLinkHandler: (String) -> Unit = { openLink(context, it, settings.linkTarget) }

    Scaffold(
        topBar = {
            TopAppBar(
                title = {
                    Text(
                        if (model.nodes.isEmpty()) "Discussion"
                        else pluralized(model.nodes.size, "comment"),
                    )
                },
                navigationIcon = {
                    IconButton(onClick = onBack) {
                        Icon(Icons.Filled.ArrowBack, contentDescription = "Back")
                    }
                },
                actions = {
                    IconButton(onClick = { model.retry(settings.autoCollapseDeepThreads) }) {
                        Icon(Icons.Filled.Refresh, contentDescription = "Refresh")
                    }
                    IconButton(onClick = { menuOpen = true }) {
                        Icon(Icons.Filled.MoreVert, contentDescription = "More")
                    }
                    DropdownMenu(expanded = menuOpen, onDismissRequest = { menuOpen = false }) {
                        story.link?.let { link ->
                            DropdownMenuItem(
                                text = { Text("Open Article") },
                                onClick = { menuOpen = false; openLinkHandler(link) },
                            )
                        }
                        DropdownMenuItem(
                            text = { Text("Open on Hacker News") },
                            onClick = { menuOpen = false; openLinkHandler(story.hnUrl) },
                        )
                        DropdownMenuItem(
                            text = {
                                Text(
                                    if (library.isSaved(story.id)) "Remove from Saved"
                                    else "Save Story",
                                )
                            },
                            onClick = { menuOpen = false; library.toggleSave(story) },
                        )
                        HorizontalDivider()
                        DropdownMenuItem(
                            text = { Text("Share Link") },
                            onClick = {
                                menuOpen = false
                                shareLink(context, story.primaryUrl, story.displayTitle)
                            },
                        )
                        DropdownMenuItem(
                            text = { Text("Share Discussion") },
                            onClick = {
                                menuOpen = false
                                shareLink(context, story.hnUrl, story.displayTitle)
                            },
                        )
                    }
                },
            )
        },
    ) { padding ->
        LazyColumn(modifier = Modifier.padding(padding)) {
            item {
                Column(
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(horizontal = 16.dp, vertical = 12.dp),
                ) {
                    Text(
                        story.displayTitle,
                        style = MaterialTheme.typography.titleLarge,
                        fontWeight = FontWeight.Bold,
                        modifier = Modifier.clickable { openLinkHandler(story.primaryUrl) },
                    )
                    story.host?.let { host ->
                        Text(
                            host,
                            style = MaterialTheme.typography.bodySmall,
                            color = MaterialTheme.colorScheme.primary,
                            modifier = Modifier.padding(top = 6.dp),
                        )
                    }
                    Row(
                        modifier = Modifier.padding(top = 10.dp),
                        horizontalArrangement = Arrangement.spacedBy(12.dp),
                        verticalAlignment = Alignment.CenterVertically,
                    ) {
                        story.score?.let {
                            Text(
                                pluralized(it, "point"),
                                style = MaterialTheme.typography.labelMedium,
                                color = MaterialTheme.colorScheme.primary,
                            )
                        }
                        Text(
                            story.author,
                            style = MaterialTheme.typography.labelMedium,
                            color = MaterialTheme.colorScheme.onSurfaceVariant,
                            modifier = Modifier.clickable { onOpenUser(story.author) },
                        )
                        Text(
                            shortAge(story.time),
                            style = MaterialTheme.typography.labelMedium,
                            color = MaterialTheme.colorScheme.onSurfaceVariant,
                        )
                    }

                    if (!story.text.isNullOrEmpty()) {
                        RichText(
                            html = story.text!!,
                            style = MaterialTheme.typography.bodyMedium,
                            modifier = Modifier.padding(top = 12.dp),
                            onLinkClick = openLinkHandler,
                        )
                    }
                }
                HorizontalDivider()

                if (model.nodes.isNotEmpty()) {
                    Row(
                        modifier = Modifier
                            .fillMaxWidth()
                            .padding(horizontal = 16.dp, vertical = 4.dp),
                        verticalAlignment = Alignment.CenterVertically,
                    ) {
                        Text(
                            pluralized(model.nodes.size, "comment").uppercase(),
                            style = MaterialTheme.typography.labelSmall,
                            color = MaterialTheme.colorScheme.onSurfaceVariant,
                            modifier = Modifier.weight(1f),
                        )
                        TextButton(
                            onClick = {
                                if (model.isFullyCollapsed) model.expandAll() else model.collapseAll()
                            },
                        ) {
                            Text(if (model.isFullyCollapsed) "Expand All" else "Collapse All")
                        }
                    }
                    HorizontalDivider()
                }
            }

            val phase = model.phase
            when {
                model.nodes.isEmpty() && phase == LoadPhase.Loading ->
                    item { LoadingState("Loading discussion…") }

                model.nodes.isEmpty() && phase is LoadPhase.Failed ->
                    item {
                        EmptyState(
                            icon = Icons.Filled.Warning,
                            title = "Something went wrong",
                            message = phase.message,
                            actionLabel = "Try Again",
                            onAction = { model.retry(settings.autoCollapseDeepThreads) },
                        )
                    }

                model.nodes.isEmpty() && phase == LoadPhase.Loaded ->
                    item {
                        EmptyState(
                            icon = Icons.Filled.List,
                            title = "No comments yet",
                            message = "Open this story on Hacker News to reply.",
                        )
                    }

                else -> items(count = visible.size) { index ->
                    val node = visible[index]
                    CommentRow(
                        node = node,
                        isCollapsed = model.isCollapsed(node.id),
                        isOriginalPoster = node.author != null && node.author == story.by,
                        onToggle = { model.toggleCollapse(node) },
                        onAuthorClick = onOpenUser,
                        onLinkClick = openLinkHandler,
                    )
                    HorizontalDivider()
                }
            }
        }
    }
}
